Source/WebCore/crypto/openssl/OpenSSLCryptoUniquePtr.h

```diff
@@ -40,134 +45,45 @@ struct OpenSSLCryptoPtrDeleter {
-    void operator()(PKCS8_PRIV_KEY_INFO* ptr) const
-    {
-        PKCS8_PRIV_KEY_INFO_free(ptr);
-    }
-};
-
-using PKCS8PrivKeyInfoPtr = OpenSSLCryptoPtr<PKCS8_PRIV_KEY_INFO>;
-
-template <>
-struct OpenSSLCryptoPtrDeleter<BIGNUM> {
-    void operator()(BIGNUM* ptr) const
-    {
-        BN_clear_free(ptr);
-    }
-};
-
-using BIGNUMPtr = OpenSSLCryptoPtr<BIGNUM>;
-
-template <>
-struct OpenSSLCryptoPtrDeleter<BN_CTX> {
-    void operator()(BN_CTX* ptr) const
-    {
-        BN_CTX_free(ptr);
-    }
-};
-
-using BNCtxPtr = OpenSSLCryptoPtr<BN_CTX>;
-
-template <>
-struct OpenSSLCryptoPtrDeleter<ECDSA_SIG> {
-    void operator()(ECDSA_SIG* ptr) const
-    {
-        ECDSA_SIG_free(ptr);
-    }
-};
-
-using ECDSASigPtr = OpenSSLCryptoPtr<ECDSA_SIG>;
-
-template <>
-struct OpenSSLCryptoPtrDeleter<ASN1_SEQUENCE_ANY> {
-    void operator()(ASN1_SEQUENCE_ANY* ptr) const
-    {
-        sk_ASN1_TYPE_pop_free(ptr, ASN1_TYPE_free);
-    }
-};
-
-using ASN1SequencePtr = OpenSSLCryptoPtr<ASN1_SEQUENCE_ANY>;
+#define DEFINE_OPENSSL_CRYPTO_PTR_FULL(alias, typeName, deleterFunc) \
+    template<> struct OpenSSLCryptoPtrDeleter<typeName> { \
+        void operator()(typeName* ptr) const { \
+            deleterFunc;                                             \
+        }                                                            \
+    };                                                               \
+    using alias = OpenSSLCryptoPtr<typeName>;
+
+#define DEFINE_OPENSSL_CRYPTO_PTR(alias, typeName, deleterFunc)      \
+    DEFINE_OPENSSL_CRYPTO_PTR_FULL(alias, typeName, deleterFunc(ptr))
+
+DEFINE_OPENSSL_CRYPTO_PTR(EvpCipherCtxPtr, EVP_CIPHER_CTX, EVP_CIPHER_CTX_free)
+DEFINE_OPENSSL_CRYPTO_PTR(EvpDigestCtxPtr, EVP_MD_CTX, EVP_MD_CTX_free)
+DEFINE_OPENSSL_CRYPTO_PTR(EvpPKeyPtr, EVP_PKEY, EVP_PKEY_free)
+DEFINE_OPENSSL_CRYPTO_PTR(EvpPKeyCtxPtr, EVP_PKEY_CTX, EVP_PKEY_CTX_free)
+
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+DEFINE_OPENSSL_CRYPTO_PTR(OsslParamBldPtr, OSSL_PARAM_BLD, OSSL_PARAM_BLD_free)
+DEFINE_OPENSSL_CRYPTO_PTR(OsslParamPtr, OSSL_PARAM, OSSL_PARAM_free)
+DEFINE_OPENSSL_CRYPTO_PTR(EVPKDFCtxPtr, EVP_KDF_CTX, EVP_KDF_CTX_free)
+DEFINE_OPENSSL_CRYPTO_PTR(EVPKDFPtr, EVP_KDF, EVP_KDF_free)
+#endif // OPENSSL_VERSION_NUMBER >= 0x30000000L
+
+// These are deprecated in OpenSSL 3. FIXME: Migrate to EvpKey. See Bug #245146.
+DEFINE_OPENSSL_CRYPTO_PTR(RSAPtr, RSA, RSA_free)
+DEFINE_OPENSSL_CRYPTO_PTR(ECKeyPtr, EC_KEY, EC_KEY_free)
+DEFINE_OPENSSL_CRYPTO_PTR(HMACCtxPtr, HMAC_CTX, HMAC_CTX_free)
+
+DEFINE_OPENSSL_CRYPTO_PTR(ECPointPtr, EC_POINT, EC_POINT_clear_free)
+DEFINE_OPENSSL_CRYPTO_PTR(PKCS8PrivKeyInfoPtr, PKCS8_PRIV_KEY_INFO, PKCS8_PRIV_KEY_INFO_free)
+DEFINE_OPENSSL_CRYPTO_PTR(BIGNUMPtr, BIGNUM, BN_clear_free)
+DEFINE_OPENSSL_CRYPTO_PTR(BNCtxPtr, BN_CTX, BN_CTX_free)
+DEFINE_OPENSSL_CRYPTO_PTR(ECDSASigPtr, ECDSA_SIG, ECDSA_SIG_free)
+DEFINE_OPENSSL_CRYPTO_PTR(X509Ptr, X509, X509_free)
+DEFINE_OPENSSL_CRYPTO_PTR(BIOPtr, BIO, BIO_free)
```

Comment:
Is there a specific plan for these last two (X509Ptr and BIOPtr) or are they just for filling out items that weren't covered before?

Reply:
They will be used in the GstWebRTC backend.